packages/api/src/external/carequality/document/process-document-query-results.ts

```diff
@@ -123,60 +81,83 @@ export async function processDocumentQueryResults({
-        const FHIRDocRef = cqToFHIR(docId, docRef, patientId);
-
-        await upsertDocumentToFHIRServer(cxId, FHIRDocRef);
-      } catch (error) {
-        console.log(`Failed to store initial doc ref in FHIR: ${errorToString(error)}`);
-        capture.error(error, {
-          extra: {
-            context: `cq.storeInitDocRefInFHIR`,
-            error,
-            docRef,
-          },
-        });
-        throw error;
-      }
+function combineDocRefs(documentQueryResults: IHEToExternalGwDocumentQuery[]): DocumentReference[] {
+  return documentQueryResults.reduce((acc: DocumentReference[], curr) => {
+    const documentReferences = curr.data.documentReference ?? [];
+
+    return [...acc, ...documentReferences];
+  }, []);
+}
+
+const getNonExistentDocRefs = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<DocumentReference[]> => {
+  const [{ existingDocRefs, nonExistingDocRefs }, FHIRDocRefs] = await Promise.all([
+    filterOutExistingDocRefsS3(documents, patientId, cxId),
+    getDocuments({ cxId, patientId }),
+  ]);
+
+  const fhirDocRefIds = new Set(FHIRDocRefs.map(d => d.id));
+
+  const foundOnStorageButNotOnFHIR = existingDocRefs.filter(f => !fhirDocRefIds.has(f.docUniqueId));
+
+  const docsToDownload = nonExistingDocRefs.concat(foundOnStorageButNotOnFHIR);
+
+  return docsToDownload;
+};
+
+type ObservedDocRefs = {
+  existingDocRefs: DocumentReference[];
+  nonExistingDocRefs: DocumentReference[];
+};
+
+const filterOutExistingDocRefsS3 = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<ObservedDocRefs> => {
+  const docIdWithExist = await Promise.allSettled(
+    documents.map(async (doc): Promise<{ docId: string; exists: boolean }> => {
+      const fileName = createDocumentFilePath(
+        cxId,
+        patientId,
+        doc.docUniqueId,
+        doc.contentType ?? undefined
+      );
+
+      const { exists } = await s3Utils.getFileInfoFromS3(fileName, s3BucketName);
+
+      return {
+        docId: doc.docUniqueId,
+        exists,
+      };
     })
   );
-}
 
-function addMetriportDocRef({
-  cxId,
-  patientId,
-  requestId,
-}: {
-  patientId: string;
-  cxId: string;
-  requestId: string;
-}) {
-  return async (document: DocumentReference): Promise<DocumentWithMetriportId> => {
-    const documentId = document.docUniqueId;
-
-    const { metriportId, originalId } = await mapDocRefToMetriport({
-      cxId,
-      patientId,
-      documentId,
-      requestId,
-      source: MedicalDataSource.CAREQUALITY,
-    });
-    return {
-      ...document,
-      originalId: originalId,
-      id: metriportId,
-    };
-  };
-}
+  const successfulDocs = docIdWithExist.flatMap(ref =>
+    ref.status === "fulfilled" && ref.value ? ref.value : []
+  );
+
+  const observedDocRefs = documents.reduce(
+    (acc: ObservedDocRefs, curr) => {
+      for (const succDoc of successfulDocs) {
+        if (succDoc.docId === curr.docUniqueId) {
```

Comment:
Do we expect to have more than one entry where the conditional is true? If not, it would be more efficient to `.find()` instead of using this loop inside the reducer (I'd use it nonetheless, as its cleaner/easier to read).